src/Database/LMDB/Simple/Internal.hs

```diff
@@ -104,31 +112,36 @@ import Foreign
   , copyBytes
   )
 
-import GHC.Exts
-  ( Constraint
-  )
-
-data ReadWrite
-data ReadOnly
-
-class Mode a where
-  isReadOnlyMode :: a -> Bool
-
-instance Mode ReadWrite where isReadOnlyMode _ = False
-instance Mode ReadOnly  where isReadOnlyMode _ = True
-
-type family SubMode a b :: Constraint where
-  SubMode a ReadWrite = a ~ ReadWrite
-  SubMode a ReadOnly  = ()
+-- | Modes of operation for 'Environment's and 'Transaction's.
+--
+-- Note: See the documentation of 'SubMode' and 'Transaction'.
+data Mode = ReadWrite | ReadOnly
+
+-- | Reflect (on the term level) whether a @'Mode'@ is read-only or read-write
+type IsMode :: Mode -> Constraint
+class IsMode mode where
+  isReadOnlyMode :: proxy mode -> Bool
+
+instance IsMode ReadWrite where
+  isReadOnlyMode _ = False
+instance IsMode ReadOnly  where
+  isReadOnlyMode _ = True
+
+-- | Both read-only and read-write transactions can run in a read-write
+-- environment, but read-only transactions can only run in read-only
+-- environments.
+type SubMode :: Mode -> Mode -> Constraint
+type family SubMode mode1 mode2 where
```

Comment:
Maybe document which mode is for the transaction and which one is for the environment?